source/h2_stream.c

```diff
@@ -87,15 +90,97 @@ struct aws_h2_stream *aws_h2_stream_new_request(
 
     /* Init H2 specific stuff */
     stream->thread_data.state = AWS_H2_STREAM_STATE_IDLE;
-    aws_linked_list_node_reset(&stream->node);
+    stream->thread_data.outgoing_message = options->request;
+    aws_http_message_acquire(stream->thread_data.outgoing_message);
 
     return stream;
 }
+
 static void s_stream_destroy(struct aws_http_stream *stream_base) {
     AWS_PRECONDITION(stream_base);
     struct aws_h2_stream *stream = AWS_CONTAINER_OF(stream_base, struct aws_h2_stream, base);
 
     AWS_H2_STREAM_LOG(DEBUG, stream, "Destroying stream");
 
+    aws_http_message_release(stream->thread_data.outgoing_message);
+
     aws_mem_release(stream->base.alloc, stream);
 }
+
+enum aws_h2_stream_state aws_h2_stream_get_state(const struct aws_h2_stream *stream) {
+    AWS_PRECONDITION_ON_CHANNEL_THREAD(stream);
+    return stream->thread_data.state;
+}
+
+static struct aws_h2_frame_headers *s_new_headers_frame(
+    struct aws_allocator *alloc,
+    const struct aws_http_message *message) {
+
+    struct aws_h2_frame_headers *headers_frame = aws_mem_calloc(alloc, 1, sizeof(struct aws_h2_frame_headers));
+    if (!headers_frame) {
+        goto error_alloc;
+    }
+
+    if (aws_h2_frame_headers_init(headers_frame, alloc)) {
+        goto error_init;
+    }
+
+    const size_t num_headers = aws_http_message_get_header_count(message);
```

Comment:
how are we handling header's size being too large for the frame here?

Reply:
Not handling it yet. I'll add an explicit TODO.

The way we encode headers needs a complete revamp to account for max frame size, and splitting the contents across CONTINUATION frames. This will probably be the next PR.